AddressRepository save method causes large memory consumption, eventually leading to out of memory crash

cherreman opened the pull request:
### Preconditions
<!--- Provide a more detailed information of environment you use -->
<!--- Magento version, tag, HEAD, etc., PHP & MySQL version, etc.. -->
1. Magento version: 2.1.2
2. PHP 7.0.15

### Steps to reproduce
1. Create a customer with a lot of addresses. In our case, we have 250+ shipping addresses.
2. Loop through all the addresses and save the data models via the address repository.

### Expected result
<!--- Tell us what should happen -->
1. Addresses are saved. No memory side effects.

### Actual result
<!--- Tell us what happens instead -->
1. With each address that is saved, the memory usage increases. In our case it is around 2mb per address.

Fix:
The fix needs to happen in two places:

The getDataModel() method in Magento\Customer\Model\Address does not take into account the default billing and shipping address that are passed. This causes the customer to be loaded, which most likely causes the memory issue.

The Magento\Customer\Model\ResourceModel\AddressRepository save() method returns the data model of the address object. When calling address->getDataModel() is does not pass in the default billing and shipping address id.

cherreman committed:
Update Address.php

Take into account default billing and shipping address id in getDataModel()

cherreman committed:
Update AddressRepository.php

Pass default billing and shipping address id to addressModel->getDataModel() in save() method.

ishakhsuvarov commented:
Hi @cherreman
Could you please look into the test fails in the linked Travis CI build? Currently it looks like majority of the jobs are failed with a fatal error.
Thank you!

cherreman committed:
Update AddressRepository.php

Check for instances of CustomerAddressModel since getDefault*Address() can also return a boolean

cherreman committed:
Update AddressRepository.php

Fixed check for instance of (too many different programming languages)

cherreman commented:
@ishakhsuvarov Oh thank you for this. The code has been changed to fix the tests. One of the test is failing with a timeout though.

maghamed commented on app/code/Magento/Customer/Model/ResourceModel/AddressRepository.php:
```diff
@@ -139,7 +139,17 @@ public function save(\Magento\Customer\Api\Data\AddressInterface $address)
 
-        return $addressModel->getDataModel();
+        $defaultBillingAddressId = null;
+        if ($customerModel->getDefaultBillingAddress() instanceof CustomerAddressModel) {
```
Magento doesn't encourage usage of instanceof operator, as this couldn't be considered as good OOP practice.
taking into account that 
`
     * @return  Address|false
     */
    public function getPrimaryAddress($attributeCode)
`
Change your conditional logic to this presentation:

`
if ($customerModel->getDefaultBillingAddress()) {
`

Also this will introduce micro optimization

cherreman replied:
@maghamed I see. What about returning Address or false? I would consider this a potential source of errors and it complicates the client code. Why not return Address or null in this case?

maghamed commented on app/code/Magento/Customer/Model/ResourceModel/AddressRepository.php:
```diff
@@ -139,7 +139,17 @@ public function save(\Magento\Customer\Api\Data\AddressInterface $address)
+        }
+
+        $defaultShippingAddressId = null;
+        if ($customerModel->getDefaultShippingAddress() instanceof CustomerAddressModel) {
```
Magento doesn't encourage usage of instanceof operator, as this couldn't be considered as good OOP practice.
taking into account that 
`
     * @return  Address|false
     */
    public function getPrimaryAddress($attributeCode)
`
Change your conditional logic to this presentation:

`
if ($customerModel->getDefaultBillingAddress()) {
`

Also this will introduce micro optimization

maghamed commented on app/code/Magento/Customer/Model/Address.php:
```diff
@@ -171,13 +171,14 @@ public function updateData(AddressInterface $address)
     {
         if ($this->getCustomerId() || $this->getParentId()) {
-            if ($this->getCustomer()->getDefaultBillingAddress()) {
+            if ((intval($defaultBillingAddressId) == 0) && $this->getCustomer()->getDefaultBillingAddress()) {
```
There is dedicated function is_null in php, but in Magento we use strict comparison equation (===) with null
So, in your case code should look like
`if ($defaultBillingAddressId === null) && $this->getCustomer()->getDefaultBillingAddress()) {`

maghamed replied:
also there is possible bug in this logic , when  $defaultBillingAddressId is equal 0 , but not null. Your assertion still would be positive

maghamed commented on app/code/Magento/Customer/Model/Address.php:
```diff
@@ -171,13 +171,14 @@ public function updateData(AddressInterface $address)
                 $defaultBillingAddressId = $this->getCustomer()->getDefaultBillingAddress()->getId();
             }
-            if ($this->getCustomer()->getDefaultShippingAddress()) {
+            if ((intval($defaultShippingAddressId) == 0) && $this->getCustomer()->getDefaultShippingAddress()) {
```
There is dedicated function is_null in php, but in Magento we use strict comparison equation (===) with null
So, in your case code should look like
`if ($defaultShippingAddressId=== null) && $this->getCustomer()->getDefaultShippingAddress()) {`

maghamed reviewed (commented):
@cherreman  Please pay attention to review comments

maghamed commented:
And Ideally to cover these changes with Unit test

cherreman committed:
removed check with instanceof for billing and shipping address

cherreman committed:
changed null check on arguments

changed check for billing and shipping address id to explicit null check

ishakhsuvarov commented:
Hi @cherreman 
Please consider covering the changes with the test. 
Thanks!

ishakhsuvarov commented:
Hi @cherreman 
I am closing this PR due to inactivity for now. Please reopen if you are going to continue work on it.
